File: client-flutter/lib/services/daily_quote_service.dart
import 'package:shared_preferences/shared_preferences.dart';
import '../models/daily_quote.dart';
import 'api_client.dart';

/// Fetches the AI-generated daily quote with two-tier caching:
///   1. In-memory (survives tab switches within a session)
///   2. SharedPreferences (survives app restarts on the same calendar day)
class DailyQuoteService {
  static DailyQuote? _memQuote;
  static String? _memDate;

  static String _today() {
    final d = DateTime.now();
    return '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
  }

  Future<DailyQuote?> fetch() async {
    final today = _today();

    // ── 1. In-memory cache ──────────────────────────────────────────────────
    if (_memQuote != null && _memDate == today) return _memQuote;

    // ── 2. SharedPreferences cache ──────────────────────────────────────────
    final prefs = await SharedPreferences.getInstance();
    if (prefs.getString('dq_date') == today) {
      final text   = prefs.getString('dq_text');
      final author = prefs.getString('dq_author');
      if (text != null && author != null && text.isNotEmpty) {
        _memQuote = DailyQuote(text: text, author: author);
        _memDate  = today;
        return _memQuote;
      }
    }

    // ── 3. API call ─────────────────────────────────────────────────────────
    final data = await apiClient.get('api/daily-quote');
    if (data is Map<String, dynamic>) {
      final q = DailyQuote.fromJson(data);
      if (q.text.isNotEmpty) {
        _memQuote = q;
        _memDate  = today;
        await prefs.setString('dq_date',   today);
        await prefs.setString('dq_text',   q.text);
        await prefs.setString('dq_author', q.author);
        return q;
      }
    }
    return null;
  }
}

File: client-ios/SihlClient/Services/DailyQuoteService.swift
import Foundation

/// Pendant zu `services/daily_quote_service.dart`.
/// Zwei-Ebenen-Cache:
///   1. In-memory (pro App-Session, überlebt Tab-Wechsel)
///   2. UserDefaults (pro Kalendertag, überlebt Neustarts)
actor DailyQuoteService {
    static let shared = DailyQuoteService()

    private var memQuote: DailyQuote?
    private var memDate: String?

    // MARK: - Public

    func fetch() async throws -> DailyQuote? {
        let today = todayString()

        // 1. In-memory
        if let q = memQuote, memDate == today { return q }

        // 2. UserDefaults
        let defaults = UserDefaults.standard
        if defaults.string(forKey: "dq_date") == today,
           let text   = defaults.string(forKey: "dq_text"),
           let author = defaults.string(forKey: "dq_author"),
           !text.isEmpty {
            let q = DailyQuote(text: text, author: author)
            store(q, date: today)
            return q
        }

        // 3. API
        let data = try await APIClient.shared.get("api/daily-quote")
        guard
            let q = try? JSONDecoder().decode(DailyQuote.self, from: data),
            !q.text.isEmpty
        else { return nil }

        store(q, date: today)
        defaults.set(today,    forKey: "dq_date")
        defaults.set(q.text,   forKey: "dq_text")
        defaults.set(q.author, forKey: "dq_author")
        return q
    }

    // MARK: - Private

    private func store(_ quote: DailyQuote, date: String) {
        memQuote = quote
        memDate  = date
    }

    private func todayString() -> String {
        let c = Calendar.current.dateComponents([.year, .month, .day], from: Date())
        return String(format: "%04d-%02d-%02d", c.year!, c.month!, c.day!)
    }
}
